napari/_qt/_qapp_model/_tests/test_processors.py

```diff
@@ -1,20 +1,108 @@
+    widget = QWidget()
+    viewer = MagicMock()
+    qtbot.addWidget(widget)
+    with pytest.raises(RuntimeError, match='No current `Viewer` found.'):
+        _add_plugin_dock_widget((widget, 'widget'))
+    _add_plugin_dock_widget((widget, 'widget'), viewer)
+
+
+def test_add_layer_data_tuples_to_viewer_invalid_data():
+    viewer = MagicMock()
+    error_data = (np.zeros((10, 10)), np.zeros((10, 10)))
+    with pytest.raises(
+        TypeError, match='Not a valid list of layer data tuples!'
+    ):
+        _add_layer_data_tuples_to_viewer(
+            data=error_data,
+            return_type=Union[ImageData, LabelsData],
+            viewer=viewer,
+        )
+
+
+def test_add_layer_data_tuples_to_viewer_valid_data(make_napari_viewer):
+    viewer = make_napari_viewer()
+    valid_data = [
+        (np.zeros((10, 10)), {'name': 'layer1'}, 'image'),
+        (np.zeros((10, 20)), {'name': 'layer1'}, 'image'),
+    ]
+    _add_layer_data_tuples_to_viewer(
+        data=valid_data,
+        return_type=Union[ImageData, LabelsData],
+        viewer=viewer,
+    )
+    assert len(viewer.layers) == 1
+    assert np.array_equal(viewer.layers[0].data, np.zeros((10, 20)))
+
+
+def test_add_layer_data_to_viewer_return_type():
     v = MagicMock()
     with pytest.raises(TypeError, match='napari supports only Optional'):
         _add_layer_data_to_viewer(
             data=np.zeros((10, 10)),
             return_type=Union[ImageData, LabelsData],
             viewer=v,
         )
+    _add_layer_data_to_viewer(
+        data=np.zeros((10, 10)),
+        return_type=Optional[ImageData],
+        viewer=v,
+    )
+
+
+def test_add_layer_data_to_viewer(make_napari_viewer):
+    viewer = make_napari_viewer()
+    _add_layer_data_to_viewer(
+        data=np.zeros((10, 10)),
+        return_type=Optional[ImageData],
+        viewer=viewer,
+        layer_name='layer1',
+    )
+    assert len(viewer.layers) == 1
+    assert np.array_equal(viewer.layers[0].data, np.zeros((10, 10)))
+    _add_layer_data_to_viewer(
+        data=np.zeros((10, 20)),
+        return_type=Optional[ImageData],
+        viewer=viewer,
+        layer_name='layer1',
+    )
+    assert len(viewer.layers) == 1
+    assert np.array_equal(viewer.layers[0].data, np.zeros((10, 20)))
+
+
+def test_add_layer_to_viewer(make_napari_viewer):
+    layer1 = Image(np.zeros((10, 10)))
+    layer2 = Image(np.zeros((10, 10)))
+    viewer = make_napari_viewer()
+    _add_layer_to_viewer(None)
+    assert len(viewer.layers) == 0
+    _add_layer_to_viewer(layer1)
+    assert len(viewer.layers) == 1
+    _add_layer_to_viewer(layer2, source={'parent': layer1}, viewer=viewer)
+    assert len(viewer.layers) == 2
+    assert layer2._source.parent == layer1
+
+
+def test_add_future_data():
+    future = MagicMock()
+    viewer = MagicMock()
+    _add_future_data(future, Union[ImageData, LabelsData])
+    _add_future_data(future, Union[ImageData, LabelsData], viewer=viewer)
```

Comment:
Check if proper mock methods are called?

Reply:
Did we expect here one call or two?

Reply:
Ah yes, the `add_done_callback` should be called twice. Will change the assert to be something like:

```python
assert future.add_done_callback.call_count == 2
```